### ISSUE TITLE
Extending build rules within a running ninja instance

### ISSUE THREAD

**Issue by elfring** [martine/ninja#892]
I am looking for a way to extend build rules while the process "ninja" is executed.
I would like to adapt the used rules for specific applications to changes from the build environment automatically. Does any [meta-programming](http://make.mad-scientist.net/the-eval-function/ "Metaprogramming Make VI — The eval Function") become feasible for better build actions here?

I imagine that [further improvements](https://github.com/martine/ninja/issues/891 "Support for built-in actions") would be needed for such a functionality with your evolving software infrastructure.

**Comment by nico** [comment]
I don't know what this means. In general, ninja's philosophy is "if it can be done by the generator, it should be done by the generator".

**Comment by elfring** [comment]
Do you find consequences from the automatic determination of available processors relevant here?

Is there a conflict with a non-goal like "build-time customization" of this software?
* The ninja program tries to start jobs in parallel by default.
* Do you want to express with any ninja construct that parallel work can only be performed after an initial data preparation step was finished?
* Does the dependency management become incomplete if several ninja runs would be needed to switch between serial and parallel work units triggered by other software tools?
